[docs] Add guide "Making a Testing Plan" (#19628)
web-platform-tests/wpt: docs/writing-tests/making-a-testing-plan.md

You may be able to perform such a search on the web. WPT is hosted on
GitHub.com, and [GitHub offers some basic functionality for querying
code](https://help.github.com/en/articles/about-searching-on-github). However,
to search effectively, you may need to use [regular
expressions](https://www.regular-expressions.info/). For that, you can
[download the WPT
repository](https://web-platform-tests.org/writing-tests/github-intro.html) and
use [git](https://git-scm.com) to perform more powerful searches.

The following table lists some common search criteria and examples of how they
can be expressed using regular expressions:

```eval_rst
================================= ================== ==========================
Criteria                          Example match      Example regular expression
================================= ================== ==========================
JavaScript identifier references  ``obj.foo()``      ``\bfoo\b``
JavaScript string literals        ``x = "foo";``     ``(["'])foo\1``
HTML tag names                    ``<foo attr>``     ``<foo(\s|>|$)``
HTML attributes                   ``<div foo=3>``    ``<[^>]+\sfoo(\s|>|=|$)``

> CSS property name                 ``style="foo: 4"`` ``([;=\"']|\s|^)foo\s+:``

Author: Ah, nice!

As described earlier in this guide, we'd certainly want to test the branch
regarding the parsing failure. However, there are many ways a string might fail
to parse--should we verify them all in the tests for `document.querySelector`?
What about `document.querySelectorAll`? Should we test them all there, too?

The answers depend on the current state of the test suite: whether or not tests
for selector parsing exist and where they are located. That's why it's best to
confer with the people who are maintaining the tests.

### Avoid excessive breadth

When the set of input values is finite, it can be tempting to test them all
exhaustively. When the set is very large, test authors can reduce repetition by
defining tests programmatically in loops.

Using advanced control flow techniques to dynamically generate tests can
actually *reduce* test quality. It may obscure the intent of the tests since
readers have to mentally "unwind" the iteration to determine what is actually
being verified. The practice is more susceptible to bugs. These bugs may not be
obvious--they may not cause failures, and they may exercise fewer cases than

> intended. Finally, tests authored using this approach often have execution time

Reviewer: I agree with @zcorpan in https://github.com/web-platform-tests/wpt/pull/19628/commits/10ae79576bcbc447bedaf7bb458537d4ac109e96#r337517178:
> I think it's easier to understand the problem by being more direct: When generating tests, it's easy to end up with a *lot* of tests which introduces the problem that running the tests takes too much time.

Author: Trying this:

> Finally, tests authored using this approach often take a relatively long time to complete, and that puts a burden on people who collect test results in large numbers.

GitHub.com, and [GitHub offers some basic functionality for querying
code](https://help.github.com/en/articles/about-searching-on-github). If your
search criteria are short and distinctive (e.g. all files containing
"querySelectorAll"), then this interface may be sufficient. However, more
complicated criteria may require [regular
expressions](https://www.regular-expressions.info/). For that, you can
[download the WPT
repository](https://web-platform-tests.org/writing-tests/github-intro.html) and
use [git](https://git-scm.com) to perform more powerful searches.

The following table lists some common search criteria and examples of how they
can be expressed using regular expressions:

```eval_rst
================================= ================== ==========================
Criteria                          Example match      Example regular expression
================================= ================== ==========================
JavaScript identifier references  ``obj.foo()``      ``\bfoo\b``
JavaScript string literals        ``x = "foo";``     ``(["'])foo\1``
HTML tag names                    ``<foo attr>``     ``<foo(\s|>|$)``

> HTML attributes                   ``<div foo=3>``    ``<[^>]+\sfoo(\s|>|=|$)``

Reviewer 2: ```suggestion
HTML attributes                   ``<div foo=3>``    ``<[a-zA-Z][^>]*\sfoo(\s|>|=|$)``
```

(This still has some false positives like `<img sizes='(min-width:500px) 500px, 100vw foo bar' srcset='x 100w, y 200w' src=x alt>` but probably good enough for most attribute names.)

> CSS property name                 ``style="foo: 4"`` ``({[;=\"']|\s|^)foo\s+:``

Reviewer 2: ```suggestion
CSS property name                 ``style="foo: 4"`` ``([{;=\"']|\s|^)foo\s*:``
```

CSS border of HTML tables. By searching through directory names and file names,
you've determined the probable location for the test: the `css/CSS2/borders/`
directory. However, there are *three hundred* files that begin with
`border-top-`! None of the names mention the `<table>` element, so any one of
the files may already be testing the case you found.

Luckily, you also know that Firefox and Chrome do not exhibit this bug. You
look at the results on [wpt.fyi](https://wpt.fyi), and you find that only three
of those tests fail in Safari but pass in Firefox and Chrome:

- `border-top-applies-to-005.xht`
- `border-top-color-applies-to-005.xht`
- `border-top-width-applies-to-005.xht`

These may not describe the behavior you're interested in testing; the only way
to know for sure is to review their contents. However, this is a much more
manageable set to work with!

### Querying file contents

Some web platform features are enabled with a predictable pattern. For example,
HTML attributes follow a fairly consistent format. If you're interested in
testing a feature like this, you may be able to learn where your tests belong
by querying the contents of the files in WPT.

> You may be able to preform such a search on the web. WPT is hosted on

Author: Good catch!

[the issues labeled with `type:missing-coverage` on
GitHub.com](https://github.com/web-platform-tests/wpt/labels/type%3Amissing-coverage).
Leave a comment if you'd like to get started with one, and don't hesitate to
ask clarifying questions!)

This guide will:

1. show you how to use the specifications to learn what kinds of tests will be
   most helpful
2. help you develop a sense for what *doesn't* need to be tested
3. demonstrate methods for figuring out which tests (if any) have already been
   written for WPT

## Understanding the "testing surface"

Web platform specifications are instructions about how a feature should work.
They're critical for implementers to "build the right thing," but they are also
important for anyone writing tests. We can use the same instructions to infer
what kinds of tests would be likely to detect mistakes. Here are a few common
patterns in specification text and the kind of tests they suggest.

### Input sources

Algorithms may accept input from many sources. Modifying the input is the only
way we can influence the browser's behavior and verify that it matches the
specifications. That's why it's helpful to be able to recognize all the

> different sources of input.

Reviewer 3: +1; I think it would be worthwhile renaming the section to something like "input and state" and talking about both from the start.

> different sources of input.

Reviewer: I was thinking two different sections, mostly so that I could link to just one of the other. (Subsections would do if there's a common discussion.)

> different sources of input.

Reviewer: My feedback is that even if they're the same if you abstract one level, you usually deal with them differently in practice. For direct input it's just a matter of permuting arguments and pretty straightforward, but for state you might need to run a series of steps, and there are non-trivial tradeoffs in whether to repeat the setup code over and over, or spawn multiple subtests from the same state to make the overall test faster.

> different sources of input.

Author: Sounds good to me. I've created a separate section for "Browser state"

> different sources of input.

Reviewer: Thanks! I haven't reviewed yet, but will add additional comments as I do.

For JavaScript APIs, the most explicit form of input is the function parameter.
For functions that are [JavaScript
methods](https://developer.mozilla.org/en-US/docs/Glossary/Method), then the
behavior may also depend on the state of the "[context
object](https://dom.spec.whatwg.org/#context-object)" (e.g. the value of `foo`
when invoking a method like `foo.method()`).

The state of the browser may also influence algorithm behavior. Examples
include the current document, the dimensions of the viewport, and the entries
in the browsing history.

> *Example* This is the first step of the `Notification` constructor from [the

Reviewer: That matches what I see, and that lack of any punctuation or box around "Example" is what I mean. Just `<i>Example:</i>` might do it.

of many different values of the *title* parameter and the *options* parameter.
The type of "the current global object" is also a form of input, so the test
suite should also include tests which execute with different types of global
objects.

### Branches

Just like traditional computer programs, specification algorithms often behave
differently based on the state of the system. And just like programs, those
differences are expressed with branching logic, usually using the same familiar
concepts of "if", "then", and "else."

When an algorithm branches based on some condition, that's an indication of an
interesting behavior that might be missed. You should write at least one test
that verifies the behavior when the branch is taken and at least one more test
that verifies the behavior when the branch is *not* taken.

*Example* The following algorithm from [the HTML
standard](https://html.spec.whatwg.org/) describes how the
`localStorage.getItem` method works:

> The `getItem`(*key*) method must return the current value associated with the
> given *key*. If the given *key* does not exist in the list associated with
> the object then this method must return null.

> This algorithm exhibits different behavior depending on whether or not an item

Reviewer 3: +1; Philip said everything I was going to, but much better :)

> This algorithm exhibits different behavior depending on whether or not an item

Reviewer 3: The wording now sgtm. I don't have any input for examples I'm afraid.

exists at the provided key. In specification language, we'd say "it branches on
the presence of the key." To test this thoroughly, we would write two tests.
One test would verify that `null` is returned when there is no item at the
provided key (that's the "branch taken" path). The other test would verify that
an item we previously stored was correctly retrieved when we called the method
with its name (that's the "branch not taken" path).

### Sequence

Even without branching, the interplay between sequential algorithm steps can
suggest interesting test cases. If two steps have observable side-effects, then
it can be useful to verify they happen in the correct order.

Most of the time, step sequence is implicit in the nature of the

> algorithm--each step operates on the result of the step that precedes it, so

Reviewer: I was just looking at the file on GitHub. If it builds correctly with sphinx, then all good.

verifying the end result implicitly verifies the sequence of the steps. But
sometimes, the order of two steps isn't particularly relevant to the result of
the overall algorithm. This makes it easier for implementations to diverge.

The most common example may be input validation. Many algorithms for JavaScript
APIs begin by verifying that the input meets some criteria. This may involve
many independent checks. The precise order of the checks may not influence

> result of the overall algorithm, but the order is well-defined and observable,

Reviewer 3: I think you're correct Philip, but I'd like to note that in `web-animations` we found many cases where Chrome violated the side-effect expectations of the specs; both in terms of step ordering and stuff like accessing a member when the spec says you can't. It seems trivial, but these things *are* observable from JS (so much is...) and it's feasible a library could be written that depends on things like accessors being called in the expected order.

> result of the overall algorithm, but the order is well-defined and observable,

Reviewer: Yes, order of many steps certainly matters, but order of throwing different types of exceptions based on different preconditions is probably close to the bottom of things that impact web developers when not interoperable, so I'd like this section to focus on the other things first.

Is there an example you could extract from the animations tests that'd work in this section?

> result of the overall algorithm, but the order is well-defined and observable,

Reviewer 2: Order of events might be more relevant, e.g.
https://bugs.chromium.org/p/chromium/issues/detail?id=1005747

Another example about order of algorithm steps: https://github.com/whatwg/html/issues/2876

> result of the overall algorithm, but the order is well-defined and observable,

Reviewer: Order of events is a great example, @zcorpan!

>    [node document](https://dom.spec.whatwg.org/#concept-node-document) set to
>    the [context object](https://dom.spec.whatwg.org/#context-object).

This algorithm begins by ensuring that the context object isn't an HTML
document. Then, it verifies that the first argument doesn't include a certain
sequence of characters. Implementations that verify those things in the
opposite order would achieve a similar result (i.e. they'd reject invalid
input), but that slight difference could cause interoperability bugs on the
web.

If we prepare input that violates *both* steps, then we can verify that they're
taken in the correct order. We can call `createCDATASection` on an HTML
document and pass the string "`]]>`"; that should produce a
`NotSupportedError`. If it produces a `InvalidCharacterError` instead, we'll
know that the implementation has incorrectly placed the second step before the
first.

## Exercising Restraint

When writing conformance tests, choosing what *not* to test is sometimes just
as hard as finding what needs testing.

### Skip ambiguity

Specifications occasionally allow browsers some discretion in how they
implement certain features. These are described using [RFC

> 2119](https://tools.ietf.org/html/rfc2119) terms like "MAY" and "OPTIONAL".

Reviewer: Thanks! Let's leave `assert_precondition` to https://github.com/web-platform-tests/wpt/issues/19844.

Even though each browser will likely maintain its own tests for the behavior it
chooses to implement, it's important that shared conformance test suites like
WPT do not include tests for any particular behavior. If the specification
doesn't require something, then neither should the tests!

*Example* The following algorithm from [the DOM
standard](https://dom.spec.whatwg.org/) powers
[`document.getElementsByTagName`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByTagName):

> > The list of elements with qualified name *qualifiedName* for a

Reviewer 3: This is a very long algorithm to put into the guide. Could we just say:

_Example_ The algorithm underpinning `document.getElementsByTagName` includes the following paragraph at the end:

> When invoked with the same argument, and as long as root's node document's type has not changed, the same HTMLCollection object may be returned as returned by an earlier call.

I don't think any of the rest of the algorithm is actually relevant to the points made?

> the same [HTMLCollection](https://dom.spec.whatwg.org/#htmlcollection) object
> may be returned as returned by an earlier call.

The final statement uses the word "may," so even though it modifies the
behavior of the algorithm, it is not appropriate to verify it in a conformance
test.

It's important to read these sections carefully, though, because the
distinction between "mandatory" behavior and "optional" behavior can be quite
nuanced. In this case, the optional behavior is never allowed if the document's
type has changed. That makes for a valid conformance test, one that verifies
browsers don't return the same result when the document's type changes.

### Don't dive too deep

Algorithms are usually composed of many other algorithms which themselves are
defined in terms of still more algorithms. It can be intimidating to consider
exhaustively testing one of those "nested" algorithms, especially when they are
shared by many different APIs.

In general, you should start by writing only "surface tests" for the nested
algorithms. That means only verifying that they exhibit the basic behavior you
are expecting.

> It's definitely important to test exhaustively, but it's just as important to

Author: Glad to hear it :)

> steps:
>
> 1. Let *s* be the result of [parse a
>    selector](https://drafts.csswg.org/selectors-4/#parse-a-selector)
>    *selectors*.
>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4)
> 2. If *s* is failure, then
>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
>    "[`SyntaxError`](https://heycam.github.io/webidl/#syntaxerror)"
>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
> 3. Return the result of [match a selector against a
>    tree](https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree)
>    with *s* and *node*'s
>    [root](https://dom.spec.whatwg.org/#concept-tree-root) using [scoping
>    root](https://drafts.csswg.org/selectors-4/#scoping-root) *node*.
>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4).

As described earlier in this guide, we'd certainly want to test the branch
regarding the parsing failure. However, there are many ways a string might fail
to parse--should we verify them all in the tests for `document.querySelector`?
What about `document.querySelectorAll`? Should we test them all there, too?

> The answers depend on the current state of the test suite: whether or not tests

Reviewer 3: I would actually be stronger here myself and say that the answer is "no". There should be separate tests for 'parse a selector', and if there aren't then the right answer is to add those, not do the testing in the `document.querySelector` test.

Reviewer 3: Those are both fair points. Ok, I am persuaded, there's more nuance here than I was allowing for :)

for selector parsing exist and where they are located. That's why it's best to
confer with the people who are maintaining the tests.

> ### Avoid excessive breadth

Reviewer: I'd suggest covering this before depth as it's a much more commonly seen problem in WPT, somewhat infamously in the editing tests.

> Many algorithms operate on input that takes the form of a scalar value. When

Reviewer: I'd skip this bit about scalar values and finite combinations, this temptation is there with strings just as much.

Reviewer 3: +1; can just remove the first sentence, and the paragraph still reads well.

the set of values is finite, it can be tempting to test them all exhaustively.
When the set is very large, test authors can reduce repetition by defining
tests programmatically in loops.

> Using advanced control flow techniques to dynamically generate tests has a

Reviewer: Does this refer to nested for loops? I disagree that this makes tests harder to understand, often it's clearer than if the same test is copy-pasted over and over. But too many subtests and time to run the tests is a problem.

I know that @domenic and @annevk also have opinions on this and especially @domenic might be more skeptical of generated tests than I am.

Reviewer 3: I am quite skeptical of generated tests. The `web-animations` test suite is *amazing*, but it uses a lot of generated tests and they cause two problems:

i. Hard to read quickly (see https://github.com/web-platform-tests/wpt/blob/master/web-animations/animation-model/animation-types/property-types.js or things like https://github.com/web-platform-tests/wpt/blob/master/web-animations/interfaces/KeyframeEffect/getKeyframes.html),

ii. Maybe worse, really hard to debug quickly. Often pulling out a single failing tests means first finding the underlying 'data-like' file, then figuring out how to comment out the parts you don't need, etc etc.

I think generated tests have value, but are a dangerous tool and can easily stray into being used for their own sake.

Reviewer: A counter example with nested for loops (`forEach` actually) that I think is *just fine*:
https://github.com/web-platform-tests/wpt/blob/master/fullscreen/api/historical.html

Unwinding these tests would not be an improvement at all, IMHO.

So, rather than pointing to "control flow techniques" as generally having drawbacks, I'd point out how it can go wrong, while being neutral on it in general or even endorsing the sensible uses.

Author: > I think it's easier to understand the problem by being more direct: When generating tests, it's easy to end up with a lot of tests which introduces the problem that running the tests takes too much time.

@zcorpan That framing doesn't account for the objection I described above. Readers may say, "of course more tests take more time; if we were optimizing for time, we wouldn't run any tests at all!" That's why I think whatever wording we choose, we have to acknowledge the "relative value" of the tests in some way.

> A counter example with nested for loops (`forEach` actually) that I think is *just fine*:
> https://github.com/web-platform-tests/wpt/blob/master/fullscreen/api/historical.html
>
> Unwinding these tests would not be an improvement at all, IMHO.

Loops may be a red herring. I get the most anxious when I see assertions or subtests inside of `if` statements. In [the DOMTokenList test I referenced earlier](https://github.com/web-platform-tests/wpt/blob/fee6738f9523f5e257accfaea8ddfef2e28cfbcd/dom/lists/DOMTokenList-coverage-for-attributes.html) the `testAttr` function is critical, and understanding what is being tested requires performing a mental transformation that reminds me of linear algebra.

But now I'm just enjoying the philosophical discussion. I don't think we should wrestle with that in this guide.

> So, rather than pointing to "control flow techniques" as generally having drawbacks, I'd point out how it can go wrong, while being neutral on it in general or even endorsing the sensible uses.

Sounds good. We were almost there, so I just softened the wording a bit. Does the latest iteration (no apologies for the pun) look good to you?

Reviewer: Those changes are in https://github.com/web-platform-tests/wpt/pull/19628/commits/f8930999036024eca2c51e848dbfefb122be167b and while it's not quite as kind to generated tests as I would be, I recognize that others have a worse experience with them, and the wording is an OK middle ground.

Reviewer: I agree with @zcorpan in https://github.com/web-platform-tests/wpt/pull/19628/commits/10ae79576bcbc447bedaf7bb458537d4ac109e96#r337517178, but will start a new thread about that specifically, as this thread will be shown as outdated by GitHub and hard to find.

number of drawbacks. It makes the test suite more difficult to understand since
readers have to mentally "unwind" the iteration to determine what is actually
being verified. It also increases the risk of bugs in the tests. These bugs may
not be obvious--they may not cause failures, and they may exercise fewer cases
than intended.

Generally speaking, such exhaustive approaches are unlikely to catch more bugs
than a handful of carefully-chosen test cases. Although the risks of dynamic

> test generation may be tolerable in some specific cases, it's usually best to

Reviewer: This is too strong of a rebuke of generated tests for me to accept :)

Reviewer 3: Despite my above dislike of generated tests, I agree with Philip. Dynamically generating `i = 1 to 10,000` is bad, but the sort of dynamic generation seen in `web-animations` is covering significantly more than 'a handful of carefully-chosen test cases' imo.

select the most interesting edge cases and move on.

*Example* We can see this consideration in the very first step of the
`Response` constructor from [the Fetch
standard](https://fetch.spec.whatwg.org/)

> The `Response`(*body*, *init*) constructor, when invoked, must run these
> steps:
>

> > 1. If *init*["`status`"] is not in the range `200` to `599`, inclusive, then

Reviewer: I don't think this is a good example because it's not the sort of case where the temptation is big and the tradeoffs hard. In fact I've never reviewed a test where someone's tried to cover each number in a range. Much more commonly is a failure to test the boundary condition and one above/below those values.

I would instead look for an example of a test that unnecessary tries too many combinations of elements and attributes "just in case" without considering what combinations could be interesting. Tests with many subtests that pass in all browsers are a good place to go looking. When I see those in review, which happens, I'd like to be able to point to this section.

Reviewer: To slightly contradict myself, I suppose it's a bad idea to use an example where there's a reasonable discussion to be had about the tradeoff, as that would distract from the point.

https://wpt.fyi/results/websockets/constructor/004.html is a bit much, but not excessive to the point that it's a problematic test.

Since I don't have an examples I think is more realistic and yet clear, I'll rescind my request :)

>    [throw](https://heycam.github.io/webidl/#dfn-throw) a `RangeError`.
>
> [...]

This function accepts exactly 400 values for the "status." With [WPT's
testharness.js](./testharness), it's easy to dynamically create one test for
each value. Unless we have reason to believe that a browser may exhibit
drastically different behavior for any of those values (e.g. correctly
accepting `546` but incorrectly rejecting `547`), then the complexity of
testing those cases probably isn't warranted.

Instead, focus on writing declarative tests for specific values which are novel
in the context of the algorithm. For ranges like in this example, testing the
boundaries is a good idea. `200` and `599` should not produce an error while
`199` and `600` should produce an error. Feel free to use what you know about
the feature to choose additional values. In this case, HTTP response status
codes are classified by the "hundred" order of magnitude, so we might also want
to test a "3xx" value and a "4xx" value.

## Assessing coverage

It's very likely that WPT already has some tests for the feature (or at least
the specification) that you're interesting in testing. In that case, you'll

> have to study what's already been done before starting to write new tests.

Reviewer: "study" seems to suggest a larger effort than is reasonable to ask for. I would "check" or "look for" tests that exist, but not spend more than 5-10 minutes to avoid test duplication.

Reviewer 3: On a meta-level, is this something that we've discussed or seen as a problem? I know as a spec author and browser implementer, I have on multiple occasions worried that I'm adding duplicate tests to WPT because finding existing tests is **so damn hard**. But maybe we've already agreed that one should just learn to love the quantum duplication?

Reviewer: "learn" sounds like an OK middle ground here.

I do think it's good to spend some time looking for existing tests, but think there are quickly diminishing returns in continuing to look when you don't find something. And have two tests you think are duplicates, there's a non-trivial risk you're wrong and they're both providing some value. Since the effort to deduplicate well is very high and the cost of duplication not very high, I don't tend to focus on it much.

Understanding the design of existing tests will let you avoid duplicating
effort, and it will also help you integrate your work more logically.

Even if the feature you're testing does *not* have any tests, you should still
keep these guidelines in mind. Sooner or later, someone else will want to
extend your work, so you ought to give them a good starting point!

### File names

The names of existing files and folders in the repository can help you find
tests that are relevant to your work. [This page on the design of
WPT](../test-suite-design) goes into detail about how files are generally laid
out in the repository.

> Generally speaking, every conformance tests is stored in a subdirectory

Reviewer: Is there some other documentation we can link to here? Something about "shortname" should be... somewhere.

Author: The W3C document is obsolete. [The link trail to a current version is a dead-end.](https://www.w3.org/pubrules) I think [this](https://www.w3.org/pubrules/doc) might be the latest, but I can't find a definition of "shortname" there.

Reviewer 2: Linking to https://web-platform-tests.org/test-suite-design.html#test-layout might be best.

Author: That document is referenced by the preceding paragraph; this one is intended as a summary. Sounds like we're good!

> The static error() method, when invoked, must run these steps:
>
> 1. Let *r* be a new [Response](https://fetch.spec.whatwg.org/#response)
>    object, whose
>    [response](https://fetch.spec.whatwg.org/#concept-response-response) is a
>    new [network error](https://fetch.spec.whatwg.org/#concept-network-error).
> 2. Set *r*'s [headers](https://fetch.spec.whatwg.org/#response-headers) to a
>    new [Headers](https://fetch.spec.whatwg.org/#headers) object whose
>    [guard](https://fetch.spec.whatwg.org/#concept-headers-guard) is
>    "`immutable`".
> 3. Return *r*.

In order to figure out where to write the test (and whether it's needed at
all), you can review the contents of the `fetch/` directory in WPT. Here's how
that looks on a UNIX-like command line:

    $ ls fetch
    api/                           data-urls/   range/
    content-encoding/              http-cache/  README.md
    content-length/                images/      redirect-navigate/
    content-type/                  metadata/    security/
    corb/                          META.yml     stale-while-revalidate/
    cors-rfc1918/                  nosniff/
    cross-origin-resource-policy/  origin/

> This test is for a behavior directly exposed through the API, so we should look

Reviewer: @annevk can confirm, but I'm pretty sure it's because there's a section in the spec by the name "API". A bunch of WHATWG specs have this but elsewhere it's rare.

Using an example from DOM involving nodes or collections might make for a more pedagogical case.

Reviewer 2: The directory structure in fetch doesn't seem to map to the sections in the spec (even within the `api` dir). I think this example is OK.

    $ ls fetch/api/response
    multi-globals/                   response-static-error.html
    response-cancel-stream.html      response-static-redirect.html
    response-clone.html              response-stream-disturbed-1.html
    response-consume-empty.html      response-stream-disturbed-2.html
    response-consume.html            response-stream-disturbed-3.html
    response-consume-stream.html     response-stream-disturbed-4.html
    response-error-from-stream.html  response-stream-disturbed-5.html
    response-error.html              response-stream-disturbed-6.html
    response-from-stream.any.js      response-stream-with-broken-then.any.js
    response-init-001.html           response-trailer.html
    response-init-002.html

There seems to be a test file for the `error` method:
`response-static-error.html`. We can open that to decide if the behavior is
already covered. If not, then we know where to [write the
test](https://github.com/web-platform-tests/wpt/pull/19601)!

### Failures on wpt.fyi

There are many behaviors that are difficult to describe in a succinct file
name. That's commonly the case with low-level rendering details of CSS

> specifications. Test authors may resort to generic number-based naming schemes

Reviewer: At the time it wasn't a last resort but by convention. It may be best to avoid this discussion here, but @frivoal might have advice on when to use this convention for new tests today.

Reviewer: I mean that this section now implicitly discourages numbered file names, and to the best of my understanding it's still a good idea in many cases today. However, if @frivoal doesn't disapprove of the framing here then I could accept it too.

Reviewer: It may amuse someone to know that only a few hours later I ran into a problem with numbered tests I hadn't considered: https://github.com/web-platform-tests/wpt/pull/8576#issuecomment-545108563

for their files, e.g. `feature-001.html`, `feature-002.html`, etc. This makes
it difficult to determine if a test case exists judging only by the names of
files.

If the behavior you want to test is demonstrated by some browsers but not by
others, you may be able to use the *results* of the tests to locate the
relevant test.

> [wpt.fyi](https://wpt.fyi) is a website which publishes results of WPT in

Reviewer: This could be folded into the above paragraph as just "on wpt.fyi". If this is the first time the reader learns about it they'll still get it from the context.

Reviewer: Very well.

various browsers. Because most browsers pass most tests, the pass/fail
characteristics of the behavior you're testing can help you filter through a
large number of highly similar tests.

> *Example* Imagine you've found a bug in the way Apple Safari renders the top

Reviewer: Omit "Apple", I think it's not part of the product name anywhere official.

CSS border of HTML tables. By searching through directory names and file names,
you've determined the probable location for the test: the `css/CSS2/borders/`
directory. However, there are *three hundred* files that begin with
`border-top-`! None of the names mention the `<table>` element, so any one of
the files may already be testing the case you found.

> Luckily, you also know that Firefox and Chrome do not exhibit this bug. You

Reviewer: Here you can link to the query syntax documentation and a permalink example that shows a concrete one.

look at the results on [wpt.fyi](https://wpt.fyi), and you find that only three
of those tests fail in Safari but pass in Firefox and Chrome:

- `border-top-applies-to-005.xht`
- `border-top-color-applies-to-005.xht`
- `border-top-width-applies-to-005.xht`

These may not describe the behavior you're interested in testing; the only way
to know for sure is to review their contents. However, this is a much more
manageable set to work with!

> ### Querying file contents

Reviewer: I think this section could be improved by very briefly pointing out how to search (GitHub or grep) and then focusing on what sorts of pattern actually work, maybe as a table with examples if GitHub and grep patterns diverge enough.

Author: Totally! I can't find anything about relevant search operators in the GitHub documentation, so I've limited the table to regular expressions. Are there any others you'd like to see covered? (I'll bet @zcorpan has ideas)

Reviewer 2: I think the list of examples is good, but maybe point out that if the thing you're searching for is a long unique word (like `querySelectorAll` or `menuitem`), then a simple substring search usually is sufficient regardless of what kind of thing it is.

Author: Good idea. Mentioning this makes our recommendation for when to use GitHub Search clearer, and it helps motivate the suggestion for regular expressions and grep.

# Making a Testing Plan

When contributing to a project as large and open-ended as WPT, it's easy to get
lost in the details. Starting with an informal testing strategy helps you
anticipate how much work will be involved, and it helps you stay focused once
you begin.

Many people come to WPT with a general testing goal in mind:

- specification authors often want to test for new spec text
- browser maintainers often want to test new features or fixes to existing
  features

> - web developers often want to test discrepancies between browsers on their web

Author: Yup, this is intended to leave the door open so such folks know that they're in the right place.

> - web developers often want to test discrepancies between browsers on their web

Reviewer: Alright, aspiration is OK :)

  applications

(If you don't have any particular goal, we can help you get started. Check out
[the issues labeled with `type:missing-coverage` on
GitHub.com](https://github.com/web-platform-tests/wpt/labels/type%3Amissing-coverage).
Leave a comment if you'd like to get started with one, and don't hesitate to
ask clarifying questions!)

This guide will:

1. show you how to use the specifications to learn what kinds of tests will be
   most helpful
2. help you develop a sense for what *doesn't* need to be tested
3. demonstrate methods for figuring out which tests (if any) have already been
   written for WPT

## Understanding the "testing surface"

> [The web platform specifications](https://platform.html5.org/) are instructions

Author: You got it. It's a shame about platform.html5.org; I really like the idea of a comprehensive index.

This guide will:

1. show you how to use the specifications to learn what kinds of tests will be
   most helpful
2. help you develop a sense for what *doesn't* need to be tested
3. demonstrate methods for figuring out which tests (if any) have already been
   written for WPT

## Understanding the "testing surface"

Web platform specifications are instructions about how a feature should work.
They're critical for implementers to "build the right thing," but they are also
important for anyone writing tests. We can use the same instructions to infer
what kinds of tests would be likely to detect mistakes. Here are a few common
patterns in specification text and the kind of tests they suggest.

### Input sources

Algorithms may accept input from many sources. Modifying the input is the most
direct way we can influence the browser's behavior and verify that it matches
the specifications. That's why it's helpful to be able to recognize all the
different sources of input.

```eval_rst
================ ==============================================================
Type of feature  Potential input sources
================ ==============================================================

> JavaScript       parameters, `context object <https://dom.spec.whatwg.org/#context-object>`_

Author: This code block is interpreted as ReStructured Text since Sphinx uses CommonMark, and CommonMark doesn't support tables. The trailing underscore is part of [the rST syntax for hyperlinks](http://docutils.sourceforge.net/docs/user/rst/quickref.html#hyperlink-targets).

> JavaScript       parameters, `context object <https://dom.spec.whatwg.org/#context-object>`_

Reviewer 3: Ah, my bad, sorry!

HTML             element content, attributes, attribute values
CSS              selector strings, property values, markup
================ ==============================================================
```

*Example:* This is the first step of the `Notification` constructor from [the
Notifications standard](https://notifications.spec.whatwg.org/#constructors):

> The Notification(title, options) constructor, when invoked, must run these steps:
>
> 1. If the [current global
>    object](https://html.spec.whatwg.org/multipage/webappapis.html#current-global-object)
>    is a
>    [ServiceWorkerGlobalScope](https://w3c.github.io/ServiceWorker/#serviceworkerglobalscope)
>    object, then [throw](https://heycam.github.io/webidl/#dfn-throw) a
>    `TypeError` exception.
> 2. Let *notification* be the result of [creating a
>    notification](https://notifications.spec.whatwg.org/#create-a-notification)
>    given *title* and *options*. Rethrow any exceptions.
>
> [...]

A thorough test suite for this constructor will include tests for the behavior
of many different values of the *title* parameter and the *options* parameter.

> Choosing those values is a challenge unto itself--see [Avoid Excessive

Reviewer 3: ```suggestion
Choosing those values is a challenge unto itself - see [Avoid Excessive
```

The '--' seems to render weirdly in the markdown; should this just be a single hyphen?

Author: [Philip also noticed this](https://github.com/web-platform-tests/wpt/pull/19628#discussion_r336737090), but since the build we're targeting (the Python Sphinx tool) automatically creates em dashes for this pattern, we're keeping it.

> Choosing those values is a challenge unto itself--see [Avoid Excessive

Reviewer 3: ```suggestion
Choosing those values can be a challenge unto itself--see [Avoid Excessive
```

> Breadth](#avoid-excessive-breadth) for advice on the topic.

Reviewer 3: ```suggestion
Breadth](#avoid-excessive-breadth) for advice.
```

### Browser state

The state of the browser may also influence algorithm behavior. Examples
include the current document, the dimensions of the viewport, and the entries
in the browsing history. Just like with direct input, a thorough set of tests
will likely need to control these values. Browser state is often more expensive
to manipulate (whether in terms of code, execution time, or system resources),
and you may want to design your tests to mitigate these costs (e.g. by writing
many subtests from the same state).

*Example:* In [the `Notification` constructor referenced

> above](https://notifications.spec.whatwg.org/#constructors), the type of "the

Reviewer 3: Because the text says 'referenced above', I clicked this link expecting it to take me back to the text up above. Instead it sent me to the notifications spec which was unexpected.

Author: If the link text said only "referenced above," then I'd understand. But because it reads "the Notification constructor referenced above," it seems clear that the target is the constructor, not the reference.

exists at the provided key. To test this thoroughly, we would write two tests:
one test would verify that `null` is returned when there is no item at the
provided key, and the other test would verify that an item we previously stored
was correctly retrieved when we called the method with its name.

### Sequence

Even without branching, the interplay between sequential algorithm steps can
suggest interesting test cases. If two steps have observable side-effects, then
it can be useful to verify they happen in the correct order.

Most of the time, step sequence is implicit in the nature of the
algorithm--each step operates on the result of the step that precedes it, so
verifying the end result implicitly verifies the sequence of the steps. But
sometimes, the order of two steps isn't particularly relevant to the result of
the overall algorithm. This makes it easier for implementations to diverge.

There are many common patterns where step sequence is observable but not
necessarily inherent to the correctness of the algorithm:

- input validation (when an algorithm verifies that two or more input values
  satisfy some criteria)
- event dispatch (when an algorithm [fires two or more

>   events](https://dom.spec.whatwg.org/#concept-event-fire))

Reviewer 3: Again, this link was a bit surprising to me; I was expecting to be sent to an example of an algorithm that fired 2+ events, not just the definition of firing an event.

Author: My main concern is that the term "fires" may not be familiar to the reader. The algorithm's full title is "fire an event," but I can see how the disparity might be confusing. I've limited the link text to just that word.

> [...]
> 4. Otherwise, if the user ended the drag-and-drop operation (e.g. by
>    releasing the mouse button in a mouse-driven drag-and-drop interface), or
>    if the `drag` event was canceled, then this will be the last iteration.
>    Run the following steps, then stop the drag-and-drop operation:
>    1. If the [current drag
>       operation](https://html.spec.whatwg.org/multipage/dnd.html#current-drag-operation)
>       is "`none`" (no drag operation) [...] Otherwise, the drag operation
>       might be a success; run these substeps:
>       1. Let *dropped* be true.
>       2. If the [current target
>          element](https://html.spec.whatwg.org/multipage/dnd.html#current-target-element)
>          is a DOM element, [fire a DND
>          event](https://html.spec.whatwg.org/multipage/dnd.html#fire-a-dnd-event)
>          named `drop` at it; otherwise, use platform-specific conventions for
>          indicating a drop.
>       3. [...]
>    2. [Fire a DND
>       event](https://html.spec.whatwg.org/multipage/dnd.html#fire-a-dnd-event)
>       named `dragend` at the [source
>       node](https://html.spec.whatwg.org/multipage/dnd.html#source-node).
>    3. [...]

> A strong test suite will verify that the `drop` event is fired as specified,

Reviewer 3: ```suggestion
A good test suite will verify that the `drop` event is fired as specified,
```

(Strong feels too... strong.)

Author: Both "strong" and "good" are pretty generic. How does "thorough" sound to you?

Reviewer 3: 👍

Reviewer 2: This is not addressed as far as I can tell.

and it will also verify that the `dragend` event is fired as specified. An even
better test suite will also verify that the `drop` event is fired *before* the
`dragend` event.

Unfortunately, there was no test like this in September of 2019. That's why
[the Chromium project was able to release a new version which changed the
sequence and interfered with real web

> applications](https://bugs.chromium.org/p/chromium/issues/detail?id=1005747).

Reviewer 3: Uh, this wording is weird. It feels like its implying that Chromium did that deliberately. How about something like:

Unfortunately, there was no test of the `drop` and `dragend` event ordering in September of 2019, when Chromium
[accidentally broke the ordering](https://bugs.chromium.org/p/chromium/issues/detail?id=1005747) with significant impact on real web applications - and didn't discover the breakage until late in their release process. This example highlights the importance of testing step sequencing!

Author: I hear what you're saying, though there are a few things about your suggestion I'd like to change. The mention of the release process is a bit distracting to me (it prompts questions about the difference between "late in the process" and "after the release"). Also, I think we can frame the conclusion more in terms of the anecdote than in terms of the guide itself.

> When making your testing plan, be sure to look carefully for event dispatch and

Reviewer 3: ```suggestion
When making your testing plan, be sure to look carefully for event dispatch ordering and
```

Author: Here and in the referenced list, the focus is on the pattern rather than the tests it implies.

Reviewer 3: Ah, I see. I had missed that it is explicitly listed as 'event dispatch' in the section above, sorry.

the other patterns listed above. They won't always be as clear as the

> abbreviated "drag" example!

Reviewer 3: ```suggestion
drag example!
```

### Optional behavior

> Specifications occasionally allow browsers some discretion in how they

Reviewer 3: ```suggestion
Specifications occasionally allow browsers discretion in how they
```

implement certain features. These are described using [RFC
2119](https://tools.ietf.org/html/rfc2119) terms like "MAY" and "OPTIONAL".

> Although browsers should not be penalized for deciding not to implement this

Reviewer 3: ```suggestion
Although browsers should not be penalized for deciding not to implement such
```

behavior, WPT offers tests that verify the correctness of the browsers which
do. Be sure to [label the test as optional according to WPT's
conventions](file-names) so that people reviewing test results know how to

> accurate interpret failures.

Reviewer 3: ```suggestion
interpret failures.
```

*Example:* The algorithm underpinning
[`document.getElementsByTagName`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByTagName)
includes the following paragraph:

> When invoked with the same argument, and as long as *root*'s [node
> document](https://dom.spec.whatwg.org/#concept-node-document)'s
> [type](https://dom.spec.whatwg.org/#concept-document-type) has not changed,
> the same [HTMLCollection](https://dom.spec.whatwg.org/#htmlcollection) object
> may be returned as returned by an earlier call.

That statement uses the word "may," so even though it modifies the behavior of
the preceding algorithm, it is strictly optional. The test we write for this
should be designated accordingly.

> It's important to read these sections carefully, though, because the

Reviewer 3: ```suggestion
It's important to read these sections carefully, because the
```

> distinction between "mandatory" behavior and "optional" behavior can be quite

Reviewer 3: ```suggestion
distinction between "mandatory" behavior and "optional" behavior can be
```

nuanced. In this case, the optional behavior is never allowed if the document's
type has changed. That makes for a mandatory test, one that verifies browsers
don't return the same result when the document's type changes.

## Exercising Restraint

When writing conformance tests, choosing what *not* to test is sometimes just
as hard as finding what needs testing.

### Don't dive too deep

> Algorithms are usually composed of many other algorithms which themselves are

Reviewer 3: ```suggestion
Algorithms are often composed of many other algorithms which themselves are
```

>
> 1. Let *s* be the result of [parse a
>    selector](https://drafts.csswg.org/selectors-4/#parse-a-selector)
>    *selectors*.
> 2. If *s* is failure, then
>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
>    "[`SyntaxError`](https://heycam.github.io/webidl/#syntaxerror)"
>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
> 3. Return the result of [match a selector against a
>    tree](https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree)
>    with *s* and *node*'s
>    [root](https://dom.spec.whatwg.org/#concept-tree-root) using [scoping
>    root](https://drafts.csswg.org/selectors-4/#scoping-root) *node*.

As described earlier in this guide, we'd certainly want to test the branch
regarding the parsing failure. However, there are many ways a string might fail
to parse--should we verify them all in the tests for `document.querySelector`?
What about `document.querySelectorAll`? Should we test them all there, too?

The answers depend on the current state of the test suite: whether or not tests
for selector parsing exist and where they are located. That's why it's best to
confer with the people who are maintaining the tests.

> ### Avoid excessive breadth

Reviewer 3: So, I'm not sure where we are on this section atm, but I think currently it reads too negatively towards dynamically generated tests. Yes, this is ironic since I previously argued that dynamic generation of tests is bad, but I think there's a balance and this text strays too far towards 'dont do it'

I would like to see it balanced slightly by acknowledging that this is a complex area, and sometimes it helps and sometimes it doesn't. I would be ok with pushing people away from it, given that this is a guide, but I just don't want to indoctrinate people into thinking never do it.

Author: I admit to being more strongly opposed to the practice than most, but I also think that by introducing a consideration without offering some advise, we may make contributors feel more doubtful than empowered. It's sort of like telling someone, "the hardest part about deep-sea fishing is making sure the robotic sharks don't destroy your boat," and moving on to discuss the intricacies of lure selection.

Really, I think we're already risking that. All but one sentence is couched in some way ("can", "often", "may"). I'm surprised that the latest revision feels too forceful. The word "sometimes" already appears in the conclusion:

"Although the risks of dynamic test generation may be tolerable in some specific cases, it’s sometimes best to select the most interesting edge cases and move on."

The risk of indoctrination from a statement like that seems pretty low.

Reviewer: The "may be tolerable in some specific cases" wording, which still remains, conveys quite strongly that this usually a very bad idea and that you'd better default to not writing for loops in your tests.

Since I now have @stephenmcgruer on my side of the argument, I'm happy to push back on this again. This, I think, is *just fine*, and better than unrolling it if the test body is longer:
```js
for (const tagName of ['audio', 'video']) {
  for (const volume of [0, 0.5, 1]) {
    test(() => {
      const element = document.createElement(tagName);
      element.volume = volume;
      assert_something_interesting();
    }, `${tagName} volume ${volume}`);
  }
}
```

Reviewer: Also, maybe advice around this makes more sense somewhere else? This guide is mostly about deciding *what* and *how much* to test, not whether to dynamically generate the tests or not.

>       might be a success; run these substeps:
>       1. Let *dropped* be true.
>       2. If the [current target
>          element](https://html.spec.whatwg.org/multipage/dnd.html#current-target-element)
>          is a DOM element, [fire a DND
>          event](https://html.spec.whatwg.org/multipage/dnd.html#fire-a-dnd-event)
>          named `drop` at it; otherwise, use platform-specific conventions for
>          indicating a drop.
>       3. [...]
>    2. [Fire a DND
>       event](https://html.spec.whatwg.org/multipage/dnd.html#fire-a-dnd-event)
>       named `dragend` at the [source
>       node](https://html.spec.whatwg.org/multipage/dnd.html#source-node).
>    3. [...]

A strong test suite will verify that the `drop` event is fired as specified,
and it will also verify that the `dragend` event is fired as specified. An even
better test suite will also verify that the `drop` event is fired *before* the
`dragend` event.

In September of 2019, [Chromium accidentally changed the ordering of the `drop`
and `dragend`
events](https://bugs.chromium.org/p/chromium/issues/detail?id=1005747), and as
a result, real web applications stopped functioning. If there had been a test

> fpr the sequence of these events, then this confusion would have been avoided.

Reviewer 3: ```suggestion
for the sequence of these events, then this confusion would have been avoided.
```

Author: How silly. Fixed

Reviewer 2: From GitHub's changed files view, this doesn't look fixed to me. Is it fixed?

### Avoid excessive breadth

When the set of input values is finite, it can be tempting to test them all
exhaustively. When the set is very large, test authors can reduce repetition by
defining tests programmatically in loops.

Using advanced control flow techniques to dynamically generate tests has a
number of drawbacks. It makes the test suite more difficult to understand since
readers have to mentally "unwind" the iteration to determine what is actually
being verified. It also increases the risk of bugs in the tests. These bugs may
not be obvious--they may not cause failures, and they may exercise fewer cases

> than intended. Finally, tests authored using this approach often have execution

Reviewer: I understand this because of the context of our discussion, but I think saying just that such tests can easily take too long to run would be clearer.

long time to complete, and that puts a burden on people who collect test
results in large numbers.

The severity of these drawbacks varies with the complexity of the generation
logic. For example, it would be pronounced in a test which conditionally made
different assertions within many nested loops. Conversely, the severity would
be low in a test which only iterated over a list of values in order to make the
same assertions about each. Recognizing when the benefits outweigh the risks

> requires discretion, so once you understand then, you should use your best

Reviewer: Understand then, typo?

Author: Good catch. Fixed!
